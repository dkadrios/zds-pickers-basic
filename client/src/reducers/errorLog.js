import { createReducer } from './utils'
import actionTypes from './actionTypes'

export const initialState = {
  currentMessage: undefined,
  logVisible: false,
  messages: [],
}

export const actions = {
  reportError: message => ({
    type: actionTypes.ERROR_OCCURRED,
    message,
  }),

  clearError: () => ({ type: actionTypes.ERROR_CLEAR }),

  clearAll: () => ({ type: actionTypes.ERROR_CLEAR_ALL }),

  showLog: logVisible => ({ type: actionTypes.ERROR_SHOW_LOG, logVisible }),
}

const handleError = (state, { message }) => {
  const messages = [...state.messages]
  const newMessage = String(message?.message ? message.message : message)

  if (messages.length && messages[messages.length - 1].text === newMessage) {
    messages[messages.length - 1].count += 1
  } else {
    messages.push({ text: newMessage, count: 1 })
  }

  return {
    ...state,
    currentMessage: newMessage,
    messages: [...messages],
  }
}

const clearError = state => ({ ...state, currentMessage: undefined })
const clearAll = state => ({
  ...state,
  currentMessage: undefined,
  logVisible: false,
  messages: [],
})

const handleShow = (state, { logVisible }) => ({
  ...state,
  logVisible,
})

export default createReducer(initialState, {
  [actionTypes.ERROR_OCCURRED]: handleError,
  [actionTypes.ERROR_CLEAR]: clearError,
  [actionTypes.ERROR_CLEAR_ALL]: clearAll,
  [actionTypes.ERROR_SHOW_LOG]: handleShow,
})
