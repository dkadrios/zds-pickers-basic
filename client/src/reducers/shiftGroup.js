/* eslint-disable no-plusplus, no-mixed-operators */
import React from 'react'
import shallowEqual from 'shallowequal'
import { statuses } from 'zds-pickers'
import { omit } from '../core/fp/objects'
import { chunk } from '../core/fp/arrays'
import InputStatusIcon from '../components/InputStatusIcon'
import { SORT_ASC, SORT_BY_ALL, SORT_ON_INPUT } from '../core/consts'
import { createReducer } from './utils'
import actionTypes from './actionTypes'
import shiftEntry from './shiftEntry'

const defaultState = {
  active: false,
  editing: false,
  editQueue: {
    entryId: -1,
    input: {
      channel: 9,
      status: 9,
      value: 0,
    },
    output: {
      channel: 9,
      status: 9,
      value: 0,
    },
  },
  /**
   * Do not confuse selectedEntryId and selectedRows
   *
   * selectedEntryId is the item being edited
   * selectedRows is used for multiple selection and drag and drop
   */
  selectedEntryId: NaN,
  selectedRows: [],
  sortOn: SORT_ON_INPUT,
  sortBy: SORT_BY_ALL,
  sortDir: SORT_ASC,
}

const modifyField = fieldName => (state, action) => ({
  ...state,
  [fieldName]: action[fieldName],
})

const addEntry = (state, action) => ({
  ...state,
  editQueue: shiftEntry({}, { ...action, channel: state.channel, entryId: -1 }),
  editing: true,
})

const removeEntry = (state, { entryId: removeEntryId }) => {
  let newState = {
    ...state,
    selectedEntryId: NaN,
    selectedRows: [],
  }

  if (Array.isArray(removeEntryId)) {
    newState = {
      ...newState,
      entries: state.entries
        .filter(({ entryId }) => !removeEntryId.includes(entryId))
        // Realign ids to indexes
        .map((entry, idx) => ({ ...entry, entryId: idx })),
    }
  } else {
    newState = {
      ...newState,
      entries: state.entries
        .filter(({ entryId }) => entryId !== removeEntryId)
        // Realign ids to indexes
        .map((entry, idx) => ({ ...entry, entryId: idx })),
    }
  }

  return newState
}

const editEntry = (state, { editEntryIdx }) => ({
  ...state,
  editQueue: state.entries.filter(({ entryId }) => entryId === editEntryIdx)[0],
  editing: true,
})

const saveEdit = (state, { editQueue }) => {
  const localEditQueue = editQueue || state.editQueue
  const { entries } = state
  const { entryId } = localEditQueue
  return {
    ...state,
    editing: false,
    selectedEntryId: entryId === -1 ? entries.length : entryId,
    selectedRows: [entryId === -1 ? entries.length : entryId],
    entries:
      entryId === -1 // Adding new entry
        ? [...entries, { ...localEditQueue, entryId: entries.length }]
        : entries // Editing existing entry
          .map(entry => (entry.entryId === entryId ? { ...localEditQueue } : entry)),
  }
}

const cancelEdit = state => ({
  ...state,
  editing: false,
})

const passThrough = (state, action) => ({
  ...state,
  editQueue: shiftEntry(state.editQueue, action),
})

const receivedGroup = (state, { entries }) => ({
  ...defaultState,
  ...state,
  entries: chunk(4)(entries).map((
    entryData,
    entryId, //
  ) => shiftEntry({}, { type: actionTypes.RECEIVED_ENTRY, entryId, entryData })),
})

const changeGroupSort = (state, { sortOn, sortBy }) => {
  const newState = {
    ...state,
    sortOn,
    sortBy,
  }

  // selectedEntryId can be NaN, which is never equal to another NaN
  const ignoreForEquivalence = omit('selectedEntryId')

  if (shallowEqual(ignoreForEquivalence(state), ignoreForEquivalence(newState))) {
    // If there was no change then assume user clicked an already sorted col.
    // Toggle the sort direction.
    newState.sortDir = 1 - newState.sortDir
  } else {
    newState.sortDir = SORT_ASC
  }
  return newState
}

const selectShiftEntry = (state, { selectedEntryId }) => ({
  ...state,
  selectedEntryId,
  selectedRows: [selectedEntryId],
})

const metaSelectEntry = (state, { meta, shift, sortedIds, idx }) => {
  let selectedRows = [...state.selectedRows]
  const entryId = sortedIds[idx]
  let i

  if (meta) {
    // Add or remove the item from selected, but keep existing items
    if (selectedRows.includes(entryId)) {
      selectedRows = selectedRows.filter(id => id !== entryId)
    } else {
      selectedRows.push(entryId)
    }
  } else if (shift && selectedRows.length === 1) {
    // If only a single item is selected, select all the items in between
    const start = sortedIds.indexOf(selectedRows[0])
    const end = sortedIds.indexOf(entryId)
    let loopEntryId
    i = start
    while (i !== end) {
      if (i > end) {
        i -= 1
      } else {
        i += 1
      }
      loopEntryId = sortedIds[i]

      /* istanbul ignore else */
      if (selectedRows.indexOf(loopEntryId) === -1) {
        selectedRows.push(loopEntryId)
      }
    }
  }

  return { ...state, selectedRows, selectedEntryId: NaN }
}

const selectAll = state => ({
  ...state,
  selectedRows: state.entries.map(({ entryId }) => entryId),
})

const selectNone = state => ({
  ...state,
  selectedRows: [],
  selectedEntryId: NaN,
})

export const gridFriendlyData = ({ input, output }) => ({
  input: {
    ...input,
    statusLabel: statuses.reduce((acc, { label, value }) => (value === input.status ? label : acc), ''),
    channel: input.channel,
  },
  output: {
    ...output,
    statusLabel: statuses.reduce((acc, { label, value }) => (value === output.status ? label : acc), ''),
    channel: output.channel,
  },
  indicator: <InputStatusIcon {...input} />,
})

export default createReducer(defaultState, {
  [actionTypes.CHANGE_GROUP_CHANNEL]: modifyField('channel'),
  [actionTypes.CHANGE_GROUP_VALUE]: modifyField('value'),
  [actionTypes.ADD_ENTRY]: addEntry,
  [actionTypes.REMOVE_ENTRY]: removeEntry,
  [actionTypes.EDIT_ENTRY]: editEntry,
  [actionTypes.SAVE_ENTRY_EDIT]: saveEdit,
  [actionTypes.CANCEL_ENTRY_EDIT]: cancelEdit,
  [actionTypes.CHANGE_ENTRY_STATUS]: passThrough,
  [actionTypes.CHANGE_ENTRY_CHANNEL]: passThrough,
  [actionTypes.CHANGE_ENTRY_VALUE]: passThrough,
  [actionTypes.RECEIVED_GROUP]: receivedGroup,
  [actionTypes.SORT_GROUP]: changeGroupSort,
  [actionTypes.SELECT_SHIFT_ENTRY]: selectShiftEntry,
  [actionTypes.SELECT_ALL_ENTRIES]: selectAll,
  [actionTypes.CLEAR_ALL_SELECTIONS]: selectNone,
  [actionTypes.ENTRY_META_CLICKED]: metaSelectEntry,
})
