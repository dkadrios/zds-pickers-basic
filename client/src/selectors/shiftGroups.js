import { createSelector } from 'reselect'
import { takeSecond } from '../core/fp/utils'
import { findObj } from '../core/fp/arrays'
import { isDisabled } from './shifter'
import { stateInputControls, stateMappings, stateShiftGroups, stateShifter, stateUser } from '.'

export const getShiftGroups = createSelector(
  stateShiftGroups,
  stateInputControls,
  (shiftGroups, inputControls) => {
    const groups = shiftGroups?.groups || []
    const decoratedGroups = groups.map((group) => {
      const assignedControls = inputControls
        .filter(({ channel, value }) => channel === group.channel && value === group.value)
        .map(({ controlId }) => controlId + 1)
        .map(controlNumber => `FS${controlNumber}`)
      return { ...group, controlLabels: assignedControls }
    })

    return {
      ...shiftGroups,
      groups: decoratedGroups,
    }
  },
)

export const getShiftGroup = createSelector(
  getShiftGroups,
  takeSecond,
  ({ groups }, groupId) => findObj('groupId', groupId)(groups),
)

export const canAddEntryToShiftGroup = createSelector(
  getShiftGroup,
  stateMappings,
  stateShifter,
  stateShiftGroups,
  stateUser,
  isDisabled,
  (shiftGroup, mappings, shifter, shiftGroups, user, disabled) => {
    const {
      errorVisible,
      exportDialogVisible,
      found,
      importDialogVisible,
      importInProcess,
      responding,
    } = shifter

    const { dialogVisible } = user

    const { editing } = shiftGroup || {}

    const {
      maxEntries,
      totalEntries,
    } = shiftGroups

    const {
      userDialogVisible,
      userImportDialogVisible,
    } = mappings

    return responding
    && found
    && !disabled
    && !editing
    && !dialogVisible
    && !importDialogVisible
    && !importInProcess
    && !userDialogVisible
    && !userImportDialogVisible
    && !exportDialogVisible
    && !errorVisible
    // && !registrationDialogVisible // TODO:
    && totalEntries < maxEntries
  },
)

export default undefined
