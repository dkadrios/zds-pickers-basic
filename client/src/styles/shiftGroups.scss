@import 'variables';
@import 'mixins';

:local {
  .container {
    border: 4px solid $color-shifter-red;
    border-radius: 0 4px 4px 4px;
  }

  .loadingCont {
    padding-top: 80px;
    text-align: center;
    font-size: 24px;
    color: $color-text-dimmer;
    height: calc(100vh - #{$grid-height} + 55px) !important;
    overflow: hidden !important;
  }
}
