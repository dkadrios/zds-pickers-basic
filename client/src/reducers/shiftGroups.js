import { assertRange } from 'zds-pickers'
import { arraySequence, chunk, splitAt } from '../core/fp/arrays'
import { sum } from '../core/fp/numbers'
import { GROUP_SIZE_BYTES, MAX_ENTRIES, MAX_GROUPS } from '../core/consts'
import { createReducer } from './utils'
import shiftGroup from './shiftGroup'
import actionTypes from './actionTypes.js'

export const actions = {
  askForGroups: () => ({
    type: actionTypes.GET_SYSEX_GROUPS,
  }),

  receivedGroups: groupData => ({
    type: actionTypes.RECEIVED_GROUPS,
    groupData,
  }),

  changeSelectedGroup: groupId => ({
    type: actionTypes.CHANGE_SELECTED_GROUP,
    groupId,
  }),

  changeGroupChannel: (groupId, channel) => ({
    type: actionTypes.CHANGE_GROUP_CHANNEL,
    groupId,
    channel: assertRange(channel, 15, 0),
  }),

  changeGroupValue: (groupId, value) => ({
    type: actionTypes.CHANGE_GROUP_VALUE,
    groupId,
    value: assertRange(value, 127, 0),
  }),

  editEntry: (groupId, editEntryIdx) => ({
    type: actionTypes.EDIT_ENTRY,
    groupId,
    editEntryIdx,
  }),

  addEntry: groupId => ({
    type: actionTypes.ADD_ENTRY,
    groupId,
  }),

  saveEntryEdit: (groupId, editQueue) => ({
    type: actionTypes.SAVE_ENTRY_EDIT,
    groupId,
    editQueue,
  }),

  cancelEntryEdit: groupId => ({
    type: actionTypes.CANCEL_ENTRY_EDIT,
    groupId,
  }),

  removeEntry: (groupId, entryId) => ({
    type: actionTypes.REMOVE_ENTRY,
    groupId,
    entryId,
  }),

  changeSort: (groupId, sortOn, sortBy) => ({
    type: actionTypes.SORT_GROUP,
    groupId,
    sortOn,
    sortBy,
  }),

  selectShiftEntry: (groupId, selectedEntryId) => ({
    type: actionTypes.SELECT_SHIFT_ENTRY,
    groupId,
    selectedEntryId,
  }),

  selectAllEntries: groupId => ({
    type: actionTypes.SELECT_ALL_ENTRIES,
    groupId,
  }),

  clearAllSelections: groupId => ({
    type: actionTypes.CLEAR_ALL_SELECTIONS,
    groupId,
  }),

  handleEntryClick: (groupId, sortedIds, idx, meta, shift) => ({
    type: actionTypes.ENTRY_META_CLICKED,
    groupId,
    sortedIds,
    idx,
    meta,
    shift,
  }),
}

const defaultState = {
  selectedGroupIdx: 0,
  maxEntries: MAX_ENTRIES,
  totalEntries: 0,
  groups: [],
}

const passThrough = (state, action) => ({
  ...state,

  groups: state.groups.map(group => (group.groupId === action.groupId ? shiftGroup(group, action) : { ...group })),
})

const changeSelectedGroup = (state, { groupId }) => ({
  ...state,
  selectedGroupIdx: groupId,
})

const addRemoveEntry = (state, action) => {
  const newState = { ...passThrough(state, action) }
  return {
    ...newState,
    totalEntries: newState.groups.reduce((acc, group) => acc + group.entries.length, 0),
  }
}

const handleShifterUnplugged = () => ({
  ...defaultState,
})

const receivedGroups = (state, { groupData }) => {
  const [preamble, rest] = splitAt(MAX_GROUPS * GROUP_SIZE_BYTES)(groupData)

  const [numEntries, channels, values] = chunk(MAX_GROUPS)(preamble)

  const groups = arraySequence(MAX_GROUPS).map((val, groupId) => ({
    groupId,
    channel: channels[groupId],
    value: values[groupId],
    label: `Group ${String.fromCharCode(65 + groupId)}`,
  }))

  return {
    ...state,
    totalEntries: sum(...numEntries),
    groups: groups.map((
      group,
      idx, //
    ) => shiftGroup(group, {
      type: actionTypes.RECEIVED_GROUP, //
      entries: rest.splice(0, numEntries[idx] * MAX_GROUPS),
    })),
  }
}

const receivedInternalState = (state, { packet }) => {
  const [numControls, ...rest] = packet
  rest.pop() // Settings flag

  const active = rest.slice(numControls * 2)

  return {
    ...state,
    groups: state.groups.map((group, idx) => ({ ...group, active: Boolean(active[idx]) })),
  }
}

export default createReducer(defaultState, {
  [actionTypes.CHANGE_SELECTED_GROUP]: changeSelectedGroup,
  [actionTypes.CHANGE_GROUP_CHANNEL]: passThrough,
  [actionTypes.CHANGE_GROUP_VALUE]: passThrough,
  [actionTypes.ADD_ENTRY]: passThrough,
  [actionTypes.EDIT_ENTRY]: passThrough,
  [actionTypes.SAVE_ENTRY_EDIT]: addRemoveEntry,
  [actionTypes.CANCEL_ENTRY_EDIT]: passThrough,
  [actionTypes.REMOVE_ENTRY]: addRemoveEntry,
  [actionTypes.CHANGE_ENTRY_STATUS]: passThrough,
  [actionTypes.CHANGE_ENTRY_CHANNEL]: passThrough,
  [actionTypes.CHANGE_ENTRY_VALUE]: passThrough,
  [actionTypes.SHIFTER_MISSING]: handleShifterUnplugged,
  [actionTypes.FACTORY_RESET]: handleShifterUnplugged,
  [actionTypes.RECEIVED_GROUPS]: receivedGroups,
  [actionTypes.SORT_GROUP]: passThrough,
  [actionTypes.SELECT_SHIFT_ENTRY]: passThrough,
  [actionTypes.RECEIVED_STATE]: receivedInternalState,
  [actionTypes.SELECT_ALL_ENTRIES]: passThrough,
  [actionTypes.CLEAR_ALL_SELECTIONS]: passThrough,
  [actionTypes.ENTRY_META_CLICKED]: passThrough,
})
