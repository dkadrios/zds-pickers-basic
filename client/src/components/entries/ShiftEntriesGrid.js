/* eslint-disable jsx-a11y/no-static-element-interactions */
import React, { useCallback } from 'react'
import PropTypes from 'prop-types'
import { assertRange } from 'zds-pickers'
import { useDispatch, useSelector } from 'react-redux'
import { entryGrid } from '../../styles/shiftGroupTable.scss'
import { compareEntry } from '../../reducers/shiftEntry'
import { delay } from '../../core/fp/utils'
import { actions } from '../../reducers/shiftGroups'
import useParamSelector from '../../hooks/useParamSelector'
import { getShiftGroup } from '../../selectors/shiftGroups'
import { isDisabled } from '../../selectors/shifter'
import { stateMappings } from '../../selectors'
import GridRow from './GridRow'

const ShiftEntriesGrid = ({ groupId }) => {
  const dispatch = useDispatch()
  const shiftGroup = useParamSelector(getShiftGroup, groupId) || {}
  const disabled = useSelector(isDisabled)
  const { channels } = useSelector(stateMappings)

  const { entries, selectedRows, sortBy, sortDir, sortOn } = shiftGroup

  const compare = compareEntry(sortOn, sortBy, sortDir)

  const sortedEntries = entries?.sort(compare) || []

  const sortedIds = sortedEntries.map(({ entryId }) => entryId)

  const selectShiftEntry = useCallback((entryId) => {
    dispatch(actions.selectShiftEntry(groupId, entryId))
  }, [dispatch, groupId])

  const selectAllEntries = useCallback(() => {
    dispatch(actions.selectAllEntries(groupId))
  }, [dispatch, groupId])

  const clearAllSelections = useCallback(() => {
    dispatch(actions.clearAllSelections(groupId))
  }, [dispatch, groupId])

  const editEntry = useCallback((editEntryIdx) => {
    dispatch(actions.editEntry(groupId, editEntryIdx))
  }, [dispatch, groupId])

  const handleEntryClick = useCallback((idx, metaKey, shiftKey) => {
    dispatch(actions.handleEntryClick(groupId, sortedIds, idx, metaKey, shiftKey))
  }, [dispatch, groupId, sortedIds])

  const saveEntryEdit = useCallback((destGroupId, editQueue) => {
    dispatch(actions.saveEntryEdit(destGroupId, editQueue))
  }, [dispatch])

  const removeEntry = useCallback((sourceGroupId, entryId) => {
    dispatch(actions.removeEntry(sourceGroupId, entryId))
  }, [dispatch])

  const handleArrowKey = (dir) => {
    if (!selectedRows.length) {
      if (sortedEntries.length) {
        selectShiftEntry(sortedEntries[0].entryId)
      }
    } else {
      const currentIndex = sortedEntries.findIndex(({ entryId }) => entryId === selectedRows[0])
      const nextIndex = assertRange(currentIndex + dir, sortedEntries.length - 1, 0)
      if (currentIndex !== nextIndex) {
        selectShiftEntry(sortedEntries[nextIndex].entryId)
      }
    }
  }

  const handleKeyDown = (event) => {
    const { key, metaKey } = event
    if (key === 'a' && metaKey) {
      selectAllEntries()
      event.preventDefault()
    }
    if (key === 'Escape') {
      clearAllSelections()
    }
    if (key === 'Enter' && selectedRows.length) {
      editEntry(selectedRows[0])
    }
    if (key === 'ArrowUp') {
      handleArrowKey(-1)
    }
    if (key === 'ArrowDown') {
      handleArrowKey(1)
    }
  }

  const handleClick = (event, idx, entryId) => {
    const { metaKey, shiftKey } = event
    if (metaKey || shiftKey) {
      event.stopPropagation()
      handleEntryClick()
    } else {
      selectShiftEntry(entryId)
    }
  }

  const handleDraggedFrom = ({ sourceGroupId, destGroupId, selectedRows: selected, copy }) => {
    const affected = entries.filter(({ entryId }) => selected.includes(entryId))

    affected.forEach(({ entryId, input, output }) => {
      saveEntryEdit(destGroupId, { input, output, entryId: -1 })
      delay(100)
      if (!copy) {
        removeEntry(sourceGroupId, entryId)
        delay(100)
      }
    })
  }

  return (
    <div
      className={entryGrid}
      onKeyDown={handleKeyDown}
      role="menuitem"
      tabIndex={0}
    >
      <div>
        {' '}
        {sortedEntries.map((entry, idx) => (
          <GridRow
            channels={channels}
            clickHandler={handleClick}
            disabled={disabled}
            editEntry={editEntry}
            entry={entry}
            groupId={groupId}
            handleDraggedFrom={handleDraggedFrom}
            idx={idx}
            key={idx}
            removeEntry={removeEntry}
            selected={selectedRows.includes(entry.entryId)}
            selectedRows={selectedRows}
          />
        ))}
      </div>
    </div>
  )
}

ShiftEntriesGrid.propTypes = { groupId: PropTypes.number.isRequired }

export default ShiftEntriesGrid
