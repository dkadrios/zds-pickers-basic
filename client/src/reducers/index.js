import { combineReducers } from 'redux'
import { reducer as midi } from 'redux-midi-fork'
import { reducer as form } from 'redux-form'
import hardwareTest from './hardwareTest'
import inputControls from './inputControls'
import mappings from './mappings'
import shifter from './shifter'
import shiftGroups from './shiftGroups'
import user from './user'
import version from './version'

const reducers = {
  form,
  hardwareTest,
  inputControls,
  mappings,
  midi,
  shifter,
  shiftGroups,
  user,
  version,
}

export default combineReducers(reducers)
