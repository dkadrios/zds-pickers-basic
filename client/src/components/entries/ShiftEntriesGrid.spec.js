import React from 'react'
// import flow from 'lodash/fp/flow'
import set from 'lodash/fp/set'
import { SORT_ASC, SORT_BY_CHANNEL, SORT_ON_INPUT } from '../../reducers/shiftGroup'
import { group } from '../../../__mocks__/shiftGroups'
import ShiftEntriesGrid from './ShiftEntriesGrid'

const { entries } = group
const groupId = 3
const editEntry = jest.fn()
const removeEntry = jest.fn()
const selectShiftEntry = jest.fn()
const selectAllEntries = jest.fn()
const clearAllSelections = jest.fn()
const handleEntryClick = jest.fn()
const saveEntryEdit = jest.fn()

const props = {
  groupId,
  editEntry,
  removeEntry,
  selectShiftEntry,
  entries,
  disabled: false,
  selectAllEntries,
  clearAllSelections,
  handleEntryClick,
  saveEntryEdit,
  sortOn: SORT_ON_INPUT,
  sortBy: SORT_BY_CHANNEL,
  sortDir: SORT_ASC,
  mappings: { channels: [] },
  selectedRows: [],
}

describe('ShiftEntriesGrid tests', () => {
  const stopPropagation = jest.fn()
  const preventDefault = jest.fn()
  const idx = 2
  const entryId = 3

  beforeEach(jest.clearAllMocks)

  it('renders correctly', () => {
    shallowExpect(ShiftEntriesGrid)(props).toMatchSnapshot()
  })

  it('clicking a row should handle register', () => {
    const event = { metaKey: false, shiftKey: true, stopPropagation }
    mount(<ShiftEntriesGrid {...props} />)
      .find('DragSource(GridRow)')
      .filter('[idx=2]')
      .instance()
      .props.clickHandler(event, idx, entryId)

    expect(stopPropagation).toHaveBeenCalled()
    expect(handleEntryClick).toHaveBeenCalledWith(groupId, [0, 1, 2, 3], idx, false, true)
  })

  it('clicking a row should handle selection', () => {
    const event = { metaKey: false, shiftKey: true, stopPropagation }
    mount(<ShiftEntriesGrid {...props} />)
      .find('DragSource(GridRow)')
      .filter('[idx=2]')
      .instance()
      .props.clickHandler(set('shiftKey', false)(event), idx, entryId)

    expect(stopPropagation).not.toHaveBeenCalled()
    expect(selectShiftEntry).toHaveBeenCalledWith(groupId, entryId)
  })

  it('pressing escape should clear all selections', () => {
    const event = {
      preventDefault,
      key: 'Escape',
      metaKey: false,
    }
    shallow(<ShiftEntriesGrid {...props} />).simulate('keydown', event)
    expect(preventDefault).not.toHaveBeenCalled()
    expect(clearAllSelections).toHaveBeenCalledWith(groupId)
  })

  it('CMD+A should select all rows', () => {
    const event = {
      preventDefault,
      key: 'a',
      metaKey: true,
    }
    shallow(<ShiftEntriesGrid {...props} />).simulate('keydown', event)
    expect(preventDefault).toHaveBeenCalled()
    expect(selectAllEntries).toHaveBeenCalledWith(groupId)
  })

  it('pressing up arrow should select the next row up', () => {
    const event = {
      preventDefault,
      key: 'ArrowUp',
      metaKey: false,
    }
    shallow(<ShiftEntriesGrid {...props} />).simulate('keydown', event)
    expect(preventDefault).not.toHaveBeenCalled()
    expect(selectShiftEntry).toHaveBeenCalledWith(groupId, 0)
  })

  it('pressing down arrow should select the next row up', () => {
    const event = {
      preventDefault,
      key: 'ArrowDown',
      metaKey: false,
    }
    shallow(<ShiftEntriesGrid {...props} />).simulate('keydown', event)
    expect(preventDefault).not.toHaveBeenCalled()
    expect(selectShiftEntry).toHaveBeenCalledWith(groupId, 0)
  })

  it('pressing enter when there are no selections should do nothing', () => {
    const event = {
      preventDefault,
      key: 'Enter',
      metaKey: true,
    }
    shallow(<ShiftEntriesGrid {...props} />).simulate('keydown', event)
    expect(editEntry).not.toHaveBeenCalled()
  })

  it('pressing enter when there is a selection should edit it', () => {
    const event = {
      preventDefault,
      key: 'Enter',
      metaKey: true,
    }
    shallow(<ShiftEntriesGrid
      {...props}
      selectedRows={[1]}
    />).simulate('keydown', event)
    expect(editEntry).toHaveBeenCalledWith(groupId, 1)
  })

  it('handleArrowKey should select the first row if none are selected', () => {
    const event = {
      preventDefault,
      key: 'ArrowUp',
      metaKey: true,
    }
    shallow(<ShiftEntriesGrid
      {...props}
      selectedRows={[]}
    />).simulate('keydown', event)
    expect(selectShiftEntry).toHaveBeenCalledWith(groupId, 0)
  })

  it('handleArrowKey should do nothing if none are selected and there are no rows', () => {
    const event = {
      preventDefault,
      key: 'ArrowUp',
      metaKey: true,
    }
    shallow(<ShiftEntriesGrid
      {...props}
      entries={[]}
      selectedRows={[]}
    />).simulate('keydown', event)
    expect(selectShiftEntry).not.toHaveBeenCalled()
  })

  it('handleArrowKey should select the next row if one is already selected', () => {
    const event = {
      preventDefault,
      key: 'ArrowDown',
      metaKey: true,
    }
    shallow(<ShiftEntriesGrid
      {...props}
      selectedRows={[1]}
    />).simulate('keydown', event)
    expect(selectShiftEntry).toHaveBeenCalledWith(groupId, 2)
  })

  it('handleArrowKey should do nothing if the next row is already the selected one', () => {
    const event = {
      preventDefault,
      key: 'ArrowUp',
      metaKey: true,
    }
    shallow(<ShiftEntriesGrid
      {...props}
      selectedRows={[0]}
    />).simulate('keydown', event)
    expect(selectShiftEntry).not.toHaveBeenCalled()
  })

  it('handleDraggedFrom should work', () => {
    const sourceGroupId = groupId
    const destGroupId = 3
    const callProps = {
      sourceGroupId,
      destGroupId,
      selectedRows: [0],
      copy: true,
    }
    mount(<ShiftEntriesGrid {...props} />)
      .find('DragSource(GridRow)')
      .filter('[idx=2]')
      .instance()
      .props.handleDraggedFrom(callProps)

    expect(saveEntryEdit).toHaveBeenCalledWith(destGroupId, {
      entryId: -1,
      input: entries[0].input,
      output: entries[0].output,
    })
    expect(removeEntry).not.toHaveBeenCalled()

    jest.resetAllMocks()

    mount(<ShiftEntriesGrid {...props} />)
      .find('DragSource(GridRow)')
      .filter('[idx=2]')
      .instance()
      .props.handleDraggedFrom(set('copy', false)(callProps))

    expect(saveEntryEdit).toHaveBeenCalledWith(destGroupId, {
      entryId: -1,
      input: entries[0].input,
      output: entries[0].output,
    })
    expect(removeEntry).toHaveBeenCalledWith(sourceGroupId, entries[0].entryId)
  })
})
