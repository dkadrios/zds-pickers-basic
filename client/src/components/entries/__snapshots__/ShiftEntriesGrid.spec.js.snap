// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShiftEntriesGrid tests renders correctly 1`] = `
<div
  onKeyDown={[Function]}
  role="menuitem"
  tabIndex={0}
>
  <div>
     
    <DragSource(GridRow)
      channels={Array []}
      clickHandler={[Function]}
      disabled={false}
      editEntry={[MockFunction]}
      entry={
        Object {
          "entryId": 0,
          "input": Object {
            "channel": 9,
            "status": 9,
            "value": 1,
          },
          "output": Object {
            "channel": 9,
            "status": 9,
            "value": 1,
          },
        }
      }
      groupId={3}
      handleDraggedFrom={[Function]}
      idx={0}
      key="0"
      removeEntry={[MockFunction]}
      selected={false}
      selectedRows={Array []}
    />
    <DragSource(GridRow)
      channels={Array []}
      clickHandler={[Function]}
      disabled={false}
      editEntry={[MockFunction]}
      entry={
        Object {
          "entryId": 1,
          "input": Object {
            "channel": 9,
            "status": 9,
            "value": 2,
          },
          "output": Object {
            "channel": 9,
            "status": 9,
            "value": 2,
          },
        }
      }
      groupId={3}
      handleDraggedFrom={[Function]}
      idx={1}
      key="1"
      removeEntry={[MockFunction]}
      selected={false}
      selectedRows={Array []}
    />
    <DragSource(GridRow)
      channels={Array []}
      clickHandler={[Function]}
      disabled={false}
      editEntry={[MockFunction]}
      entry={
        Object {
          "entryId": 2,
          "input": Object {
            "channel": 9,
            "status": 9,
            "value": 3,
          },
          "output": Object {
            "channel": 9,
            "status": 9,
            "value": 3,
          },
        }
      }
      groupId={3}
      handleDraggedFrom={[Function]}
      idx={2}
      key="2"
      removeEntry={[MockFunction]}
      selected={false}
      selectedRows={Array []}
    />
    <DragSource(GridRow)
      channels={Array []}
      clickHandler={[Function]}
      disabled={false}
      editEntry={[MockFunction]}
      entry={
        Object {
          "entryId": 3,
          "input": Object {
            "channel": 9,
            "status": 9,
            "value": 4,
          },
          "output": Object {
            "channel": 9,
            "status": 9,
            "value": 4,
          },
        }
      }
      groupId={3}
      handleDraggedFrom={[Function]}
      idx={3}
      key="3"
      removeEntry={[MockFunction]}
      selected={false}
      selectedRows={Array []}
    />
  </div>
</div>
`;
